generator/src/main.rs

```diff
@@ -192,27 +192,112 @@ impl Parser {
         }
     }
 
-    fn parse_feature_required(&mut self) {
+    fn parse_extensions(&mut self) {
+        loop {
+            use XmlEvent::*;
+            match self.reader.next().expect("failed to parse XML") {
+                StartElement {
+                    name, attributes, ..
+                } => match &name.local_name[..] {
+                    "extension" => {
+                        if let StartElement { name, .. } =
+                            self.reader.next().expect("failed to parse XML")
+                        {
+                            if name.local_name == "require" {
+                                self.parse_extension_required(&attributes);
+                            } else {
+                                eprintln!("unexpected extension element: {}", name.local_name);
+                            }
+                        } else {
+                            eprintln!("unexpected extension event");
+                        }
+                        self.finish_element();
+                    }
+                    _ => {
+                        eprintln!("unimplemented extensions element: {}", name.local_name);
+                        self.finish_element();
+                    }
+                },
+                EndElement { name } => {
+                    if name.local_name == "extensions" {
+                        break;
+                    }
+                    eprintln!("unexpected end element: {}", name);
+                }
+                EndDocument => {
+                    panic!("unexpected end of document");
+                }
+                _ => {}
+            }
+        }
+    }
+
+    fn parse_extension_required(&mut self, attrs: &[OwnedAttribute]) {
+        let ext_name = Rc::<str>::from(attr(attrs, "name").unwrap());
+        let ext_number = attr(attrs, "number").unwrap().parse::<i32>().unwrap();
+        let (ext_version, commands) = self.parse_required(Some(&ext_name), Some(ext_number));
+        if attr(attrs, "supported").map_or(false, |x| x == "disabled") {
+            self.disabled_exts.insert(ext_name);
+        } else {
+            let provisional = attr(attrs, "provisional").map_or(false, |x| x == "true");
+
+            let (tag_name, _) = split_ext_tag(&ext_name);
+            let ext = Extension {
+                name: ext_name.clone(),
+                version: ext_version.unwrap(),
+                commands,
+            };
+
+            if let Some(tag) = self.extensions.get_mut(tag_name) {
+                tag.extensions.push(ext);
+            } else if provisional {
+                self.extensions.insert(
+                    tag_name.into(),
+                    Tag {
+                        extensions: vec![ext],
+                    },
+                );
+            } else {
+                eprintln!("ignoring extension with unlisted tag: {}", ext_name);
+            }
+        }
+    }
+
+    fn parse_required(
```

Comment:
Can you squash this into the commit that introduces feature parsing? Or better yet, have an initial commit that factors out the shared code, followed by one that introduces feature parsing? That would make review much easier.

Reply:
Done!